src/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.cs

```diff
@@ -79,33 +85,18 @@ public IEnumerable<StateSet> GetStateSets(Project project)
             /// <summary>
             /// Return <see cref="StateSet"/>s for the given <see cref="Project"/>. 
             /// This will either return already created <see cref="StateSet"/>s for the specific snapshot of <see cref="Project"/> or
-            /// It will create new <see cref="StateSet"/>s for the <see cref="Project"/> and update internal state.
-            /// 
-            /// since this has a side-effect, this should never be called concurrently. and incremental analyzer (solution crawler) should guarantee that.
+            /// it will create new <see cref="StateSet"/>s for the <see cref="Project"/> and update internal state.
             /// </summary>
-            public ImmutableArray<StateSet> GetOrUpdateStateSets(Project project)
-            {
-                var projectStateSets = GetOrUpdateProjectStateSets(project);
-                return GetOrCreateHostStateSets(project, projectStateSets).OrderedStateSets.AddRange(projectStateSets.StateSetMap.Values);
-            }
-
-            /// <summary>
-            /// Return <see cref="StateSet"/>s for the given <see cref="Project"/>. 
-            /// This will either return already created <see cref="StateSet"/>s for the specific snapshot of <see cref="Project"/> or
-            /// It will create new <see cref="StateSet"/>s for the <see cref="Project"/>.
-            /// Unlike <see cref="GetOrUpdateStateSets(Project)"/>, this has no side effect.
-            /// </summary>
-            public IEnumerable<StateSet> GetOrCreateStateSets(Project project)
+            public ImmutableArray<StateSet> GetOrCreateStateSets(Project project)
```

Comment:
i wonder if it would be cleaner with less allocs to have this return two arrays (the host state sets, and the project state sets) and keep them separate.  (not for thsi pr, just something i was curious about).